text/3716-target-modifiers.md

```diff
@@ -0,0 +1,523 @@
+currently limited to Hard- vs Soft-Float which can cause issues when linked.
+However for RISC-V targets, the F (32-bit hardware float), D (64-bit hardware
+float), and Q (128-bit hardware float) extensions all can [potentially change
+the ABI][riscv-float], which would increase the number of required targets. The
+[E extension][riscv-e] may also change the ABI by limiting the number of
+registers used by the I (integer operations) extension.
+
+[riscv-float]: https://github.com/riscv-non-isa/riscv-elf-psabi-doc/blob/master/riscv-cc.adoc#named-abis
+[riscv-e]: https://github.com/riscv/riscv-isa-manual/blob/main/src/rv32e.adoc
+
+## Existing -C flags that are unsound
+
+It has recently been discovered that several existing `-C` flags modify the
+ABI, making them unsound. Examples:
+
+* [`-Csoft-float`](https://github.com/rust-lang/rust/issues/129893)
+* [`-Ctarget-feature=-neon`](https://github.com/rust-lang/rust/issues/131058)
+* [`-Clinker-plugin-lto`](https://github.com/rust-lang/rust/issues/127979)
+* [`-Cllvm-args`](https://github.com/rust-lang/rust/issues/131800#issuecomment-2418595757)
+* Possibly `-Ccode-model` and `-Crelocation-model`
+
+This problem is a new discovery and it's still not clear how to solve it.
+Target modifiers will not solve all of the flags; [some flags are just
+unfixable and need to be removed][issue130968]. But I expect that target
+modifiers will be the solution for some of these flags.
+
+[issue130968]: https://github.com/rust-lang/rust/issues/130968
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The Rust compiler has many flags that affect how source code is turned into
+machine code. Some flags can be turned on and off for each CU (compilation
+unit) in your project separately, and other flags must be applied to the entire
+application as a whole. The typical reason for flags to be in the latter
+category is that change some aspect of the ABI. For example,
+`-Zreg-struct-return` changes how to return a struct from a function call, and
+both the caller and callee must agree on how to do that even if they are in
+different CUs.
+
+The Rust compiler will detect if you incorrectly use a flag that must be
+applied to the application as a whole. For example, if you compile the standard
+library with `-Zreg-struct-return`, but don't pass the flag when compiling a
+dependency, then you will get the following error:
+```
+error: mixing -Zreg-struct-return will cause an ABI mismatch
+
+help: This error occurs because the -Zreg-struct-return flag modifies the ABI, 
+      and different crates in your project were compiled with inconsistent
+      settings.
+help: To resolve this, ensure that -Zreg-struct-return is set to the same value
+      for all crates during compilation.
+help: To ignore this error, recompile with the following flag:
+      -Cunsafe-allow-abi-mismatch=reg-struct-return
+```
+As an escape hatch, you can use `-Cunsafe-allow-abi-mismatch=reg-struct-return`
```

Comment:
There's an in-progress implementation of this here - rust-lang/rust#133138 - it implements this as a lint (so that you could `#[allow(..)]` this in a crate) and allows opting-out of all mismatch errors by setting this flag to an empty value. Both of these decisions seem undesirable to me, but wanted to ask here what you intended with the RFC.

Reply:
Ah, thanks for sharing that. I did not know about that PR.

It should not be a lint. This is for the same reason that [“Unsafe code was used outside of an unsafe block”](https://doc.rust-lang.org/error_codes/E0133.html) is not a lint.